pkg/router/controller/ingress.go

```diff
@@ -0,0 +1,550 @@
+		})
+	}
+
+	return routeEvents
+}
+
+// cacheTLS attempts to populate the tls cache for all secrets referenced by an ingress.
+func (it *IngressTranslator) cacheTLS(ingressTLS []extensions.IngressTLS, namespace, ingressKey string) bool {
+	success := true
+
+	for _, tls := range ingressTLS {
+		secretKey := getKey(namespace, tls.SecretName)
+
+		var refTLS *referencedTLS
+
+		if refTLS = it.tlsMap[secretKey]; refTLS == nil {
+			// Attempt to retrieve a secret the first time it is referenced
+
+			refTLS = &referencedTLS{
+				ingressKeys: sets.String{},
+			}
+
+			it.tlsMap[secretKey] = refTLS
+
+			// Attempt to retrieve the secret and read tls configuration from it.
+			// If retrieval or reading fails, the tls for the cache entry will be
+			// nil and only be populated in response to a subsequent event for the
+			// secret that provides valid tls data.
+			//
+			// TODO should initial retrieval be retried to minimize the chances
+			// that a temporary connectivity failure delays route availability
+			// until the next sync event (when the secret will next be seen)?
+			if secret, err := it.client.Secrets(namespace).Get(tls.SecretName); err != nil {
+				glog.V(4).Infof("Error retrieving secret %v: %v", secretKey, err)
+			} else {
+				newTLS, err := tlsFromSecret(secret)
+				// If an error was encountered, newTLS will be nil
+				if err != nil {
+					glog.V(4).Info(err)
+				}
+				refTLS.tls = newTLS
+			}
+		}
+
+		// Indicate that the secret is referenced by this ingress
+		refTLS.ingressKeys.Insert(ingressKey)
+
+		if refTLS.tls == nil {
+			glog.V(4).Infof("Unable to source TLS configuration for ingress %v from secret %v", ingressKey, secretKey)
+			success = false
+		}
+	}
+
+	return success
+}
+
+// dereferenceIngress ensures the given ingress is removed from the cache.
+func (it *IngressTranslator) dereferenceIngress(ingress *extensions.Ingress) {
+	if ingress != nil {
+		key := api.GetResourceKey(ingress.ObjectMeta)
+		delete(it.ingressMap, key)
+		it.dereferenceTLS(ingress.Spec.TLS, ingress.Namespace, key)
+	}
+}
+
+// dereferenceTLS removes references from ingress tls to cached tls.
+func (it *IngressTranslator) dereferenceTLS(ingressTLS []extensions.IngressTLS, namespace, ingressKey string) {
+	for _, tls := range ingressTLS {
+		secretKey := getKey(namespace, tls.SecretName)
+		if refTLS, ok := it.tlsMap[secretKey]; ok {
+			refTLS.ingressKeys.Delete(ingressKey)
+
+			// TLS that is no longer referenced can be deleted
+			if len(refTLS.ingressKeys) == 0 {
+				delete(it.tlsMap, secretKey)
+			}
+		}
+	}
+}
+
+// tlsForHost attempts to retrieve ingress tls configuration for the given host
+func (it *IngressTranslator) tlsForHost(ingressTLS []extensions.IngressTLS, namespace, host string) *cachedTLS {
+	for _, tls := range ingressTLS {
+		// TODO what if a host is matched by more than one IngressTLS instance?
+
+		// TODO what if IngressTLS doesn't specify Hosts?  There's mention of defaulting to a controller-specific
```

Comment:
By 'the default', do you mean the generated host name?  Currently the host name is generated by the unique_host plugin, so its not currently available when the ingress events are being generated.  Should the ingress translator be generating the host name?

I think this is problematic either way.  If an ingress is created with rules that lack a host name and tls that requires a specific host name to be applied, the hostname template for a given router could mean the difference between tls being applied or not being applied.  Not sure there's any way to solve this, maybe the best that can be done is logging when ingress paths will not have tls applied because no host matches.